File: Linear/singly-linked-list/python/singly_linked_list.py
# Author: Christopher Forte
# Date: March 14, 2026
# Title: Singly Linked List
# Description: Personal take on a linear chain of singly-linked nodes supporting O(1) front operations and O(n) indexed traversal.

from typing import TypeVar, Generic, Optional

T = TypeVar("T")


class SinglyLinkedList(Generic[T]):

    class _Node:
        __slots__ = ("data", "next")

        def __init__(self, data: object) -> None:
            self.data = data
            self.next: Optional["SinglyLinkedList._Node"] = None

    def __init__(self) -> None:
        # Initializes an empty list with no nodes.
        # Input:  None
        # Output: None
        self._head: Optional[SinglyLinkedList._Node] = None
        self._tail: Optional[SinglyLinkedList._Node] = None
        self._length: int = 0

    def add_first(self, value: T) -> None:
        # Prepends a new value before the current head.
        # Input:  value (T)
        # Output: None
        node = self._Node(value)
        node.next = self._head
        self._head = node
        if self._tail is None:
            self._tail = self._head
        self._length += 1

    def add_last(self, value: T) -> None:
        # Appends a new value after the current tail.
        # Input:  value (T)
        # Output: None
        node = self._Node(value)
        if self._tail:
            self._tail.next = node
        self._tail = node
        if self._head is None:
            self._head = self._tail
        self._length += 1

    def add_at(self, index: int, value: T) -> None:
        # Inserts a new value at the specified zero-based index.
        # Input:  index (int), value (T)
        # Output: None
        if index < 0 or index > self._length:
            raise IndexError("Index out of bounds")
        if index == 0:
            self.add_first(value)
            return
        if index == self._length:
            self.add_last(value)
            return
        prev = self._head
        for _ in range(index - 1):
            prev = prev.next
        node = self._Node(value)
        node.next = prev.next
        prev.next = node
        self._length += 1

    def remove_first(self) -> T:
        # Removes and returns the first element.
        # Input:  None
        # Output: T
        if not self._head:
            raise IndexError("List is empty")
        val = self._head.data
        self._head = self._head.next
        if self._head is None:
            self._tail = None
        self._length -= 1
        return val

    def remove_last(self) -> T:
        # Traverses to the second-to-last node and removes the tail.
        # Input:  None
        # Output: T
        if not self._head:
            raise IndexError("List is empty")
        if self._head is self._tail:
            return self.remove_first()
        prev = self._head
        while prev.next is not self._tail:
            prev = prev.next
        val = self._tail.data
        self._tail = prev
        self._tail.next = None
        self._length -= 1
        return val

    def remove_at(self, index: int) -> T:
        # Removes and returns the element at the specified index.
        # Input:  index (int)
        # Output: T
        if index < 0 or index >= self._length:
            raise IndexError("Index out of bounds")
        if index == 0:
            return self.remove_first()
        if index == self._length - 1:
            return self.remove_last()
        prev = self._head
        for _ in range(index - 1):
            prev = prev.next
        target = prev.next
        val = target.data
        prev.next = target.next
        self._length -= 1
        return val

    def get(self, index: int) -> T:
        # Returns the element at the specified index by linear traversal.
        # Input:  index (int)
        # Output: T
        if index < 0 or index >= self._length:
            raise IndexError("Index out of bounds")
        curr = self._head
        for _ in range(index):
            curr = curr.next
        return curr.data

    def first(self) -> T:
        # Returns the head element without removing it.
        # Input:  None
        # Output: T
        if not self._head:
            raise IndexError("List is empty")
        return self._head.data

    def last(self) -> T:
        # Returns the tail element without removing it.
        # Input:  None
        # Output: T
        if not self._tail:
            raise IndexError("List is empty")
        return self._tail.data

    def contains(self, value: T) -> bool:
        # Returns True if the list contains the given value.
        # Input:  value (T)
        # Output: bool
        curr = self._head
        while curr:
            if curr.data == value:
                return True
            curr = curr.next
        return False

    def reverse(self) -> None:
        # Reverses all nodes in place using iterative pointer reassignment.
        # Input:  None
        # Output: None
        prev = None
        curr = self._head
        self._tail = self._head
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        self._head = prev

    def clear(self) -> None:
        # Resets the list to empty by unlinking all nodes.
        # Input:  None
        # Output: None
        self._head = None
        self._tail = None
        self._length = 0

    def size(self) -> int:
        # Returns the number of elements in the list.
        # Input:  None
        # Output: int
        return self._length

    def is_empty(self) -> bool:
        # Returns True if the list holds no elements.
        # Input:  None
        # Output: bool
        return self._length == 0

    def __len__(self) -> int:
        # Returns the number of elements for use with len().
        # Input:  None
        # Output: int
        return self._length

    def __repr__(self) -> str:
        # Returns a string of elements separated by arrows.
        # Input:  None
        # Output: str
        parts = []
        curr = self._head
        while curr:
            parts.append(str(curr.data))
            curr = curr.next
        return " -> ".join(parts) + " -> null"

File: Linear/singly-linked-list/typescript/SinglyLinkedList.ts
/*
 * Author: Christopher Forte
 * Date: March 14, 2026
 * Title: Singly Linked List
 * Description: Personal take on a linear chain of singly-linked nodes supporting O(1) front operations and O(n) indexed traversal.
 */

interface SLLNode<T> {
  data: T;
  next: SLLNode<T> | null;
}

export class SinglyLinkedList<T> {
  private head: SLLNode<T> | null = null;
  private tail: SLLNode<T> | null = null;
  private _length: number = 0;

  // Prepends a new value before the current head.
  // Input:  value (T)
  // Output: void
  addFirst(value: T): void {
    const node: SLLNode<T> = { data: value, next: this.head };
    this.head = node;
    if (!this.tail) this.tail = this.head;
    this._length++;
  }

  // Appends a new value after the current tail.
  // Input:  value (T)
  // Output: void
  addLast(value: T): void {
    const node: SLLNode<T> = { data: value, next: null };
    if (this.tail) this.tail.next = node;
    this.tail = node;
    if (!this.head) this.head = this.tail;
    this._length++;
  }

  // Inserts a new value at the specified zero-based index.
  // Input:  index (number), value (T)
  // Output: void
  addAt(index: number, value: T): void {
    if (index < 0 || index > this._length) throw new RangeError("Index out of bounds");
    if (index === 0) return this.addFirst(value);
    if (index === this._length) return this.addLast(value);
    let prev = this.head!;
    for (let i = 0; i < index - 1; i++) prev = prev.next!;
    const node: SLLNode<T> = { data: value, next: prev.next };
    prev.next = node;
    this._length++;
  }

  // Removes and returns the first element.
  // Input:  none
  // Output: T
  removeFirst(): T {
    if (!this.head) throw new RangeError("List is empty");
    const val = this.head.data;
    this.head = this.head.next;
    if (!this.head) this.tail = null;
    this._length--;
    return val;
  }

  // Traverses to the second-to-last node and removes the tail.
  // Input:  none
  // Output: T
  removeLast(): T {
    if (!this.head) throw new RangeError("List is empty");
    if (this.head === this.tail) return this.removeFirst();
    let prev = this.head;
    while (prev.next !== this.tail) prev = prev.next!;
    const val = this.tail!.data;
    this.tail = prev;
    this.tail.next = null;
    this._length--;
    return val;
  }

  // Removes and returns the element at the specified index.
  // Input:  index (number)
  // Output: T
  removeAt(index: number): T {
    if (index < 0 || index >= this._length) throw new RangeError("Index out of bounds");
    if (index === 0) return this.removeFirst();
    if (index === this._length - 1) return this.removeLast();
    let prev = this.head!;
    for (let i = 0; i < index - 1; i++) prev = prev.next!;
    const target = prev.next!;
    prev.next = target.next;
    this._length--;
    return target.data;
  }

  // Returns the element at the specified index by linear traversal.
  // Input:  index (number)
  // Output: T
  get(index: number): T {
    if (index < 0 || index >= this._length) throw new RangeError("Index out of bounds");
    let curr = this.head!;
    for (let i = 0; i < index; i++) curr = curr.next!;
    return curr.data;
  }

  // Returns the head element without removing it.
  // Input:  none
  // Output: T
  first(): T {
    if (!this.head) throw new RangeError("List is empty");
    return this.head.data;
  }

  // Returns the tail element without removing it.
  // Input:  none
  // Output: T
  last(): T {
    if (!this.tail) throw new RangeError("List is empty");
    return this.tail.data;
  }

  // Returns true if the list contains the given value.
  // Input:  value (T)
  // Output: boolean
  contains(value: T): boolean {
    let curr = this.head;
    while (curr) {
      if (curr.data === value) return true;
      curr = curr.next;
    }
    return false;
  }

  // Reverses all nodes in place using iterative pointer reassignment.
  // Input:  none
  // Output: void
  reverse(): void {
    let prev: SLLNode<T> | null = null;
    let curr: SLLNode<T> | null = this.head;
    this.tail = this.head;
    while (curr) {
      const next = curr.next;
      curr.next = prev;
      prev = curr;
      curr = next;
    }
    this.head = prev;
  }

  // Resets the list to empty by unlinking all nodes.
  // Input:  none
  // Output: void
  clear(): void { this.head = null; this.tail = null; this._length = 0; }

  // Returns the number of elements in the list.
  // Input:  none
  // Output: number
  size(): number { return this._length; }

  // Returns true if the list holds no elements.
  // Input:  none
  // Output: boolean
  isEmpty(): boolean { return this._length === 0; }

  // Returns a string of elements separated by arrows.
  // Input:  none
  // Output: string
  toString(): string {
    const parts: string[] = [];
    let curr = this.head;
    while (curr) { parts.push(String(curr.data)); curr = curr.next; }
    return parts.join(" -> ") + " -> null";
  }
}
